gemfirexd/tools/src/dunit/java/com/pivotal/gemfirexd/transactions/SnapshotTxGIIDUnit.java

```diff
@@ -0,0 +1,399 @@
+    AttributesFactory attr = new AttributesFactory();
+    attr.setPartitionAttributes(prAttr);
+    attr.setDataPolicy(DataPolicy.PERSISTENT_PARTITION);
+    attr.setDiskStoreName("disk");
+    attr.setDiskSynchronous(true);
+    attr.setConcurrencyChecksEnabled(true);
+
+    Region pr = cache.createRegion(partitionedRegionName, attr.create());
+    assertNotNull(pr);
+  }
+
+  public static void createRR(String regionName) {
+    final GemFireCacheImpl cache = GemFireCacheImpl.getInstance();
+    assertNotNull(cache);
+    DiskStore ds = cache.findDiskStore("disk");
+    if (ds == null) {
+      ds = cache.createDiskStoreFactory()
+          .setDiskDirs(getDiskDirs()).create("disk");
+    }
+    AttributesFactory attr = new AttributesFactory();
+    attr.setConcurrencyChecksEnabled(true);
+    attr.setDataPolicy(DataPolicy.PERSISTENT_REPLICATE);
+
+
+    Region rr = cache.createRegion(regionName, attr.create());
+    assertNotNull(rr);
+  }
+
+  private void sleep(long millis) {
+    try {
+      Thread.sleep(millis);
+    } catch (InterruptedException e) {
+      e.printStackTrace();
+    }
+  }
+
+  private void putSomeValues(int numValues) {
+    final GemFireCacheImpl cache = GemFireCacheImpl.getInstance();
+    final Region r = cache.getRegion(regionName);
+    r.getCache().getCacheTransactionManager().begin(IsolationLevel.SNAPSHOT, null);
+    Map<Integer, Integer> m = new HashMap();
+    for (int i = 0; i < numValues; i++) {
+      r.put(i, i);
+    }
+    r.getCache().getCacheTransactionManager().commit();
+  }
+
+  public void testSnapshotGII_concurrentWrites() throws Exception {
+
+    startVMs(0, 2);
+    Properties props = new Properties();
+    final Connection conn = TestUtil.getConnection(props);
+    conn.createStatement();
+
+    VM server1 = this.serverVMs.get(0);
+    VM server2 = this.serverVMs.get(1);
+    stopVMNums(-2);
+
+    server1.invoke(SnapshotTxGIIDUnit.class, "createPR", new Object[]{regionName, 2, 1});
+
+    // Put some values to initialize buckets
+    server1.invoke(new SerializableRunnable() {
+      @Override
+      public void run() {
+        putSomeValues(500);
+      }
+    });
+
+    blockGII(-2, InitialImageOperation.GIITestHookType.BeforeRequestRVV);
+
+    restartServerVMNums(new int[]{2}, 0, null, null);
+
+    server2 = this.serverVMs.get(1);
+
+    AsyncInvocation ai2 = server1.invokeAsync(new SerializableRunnable() {
+      @Override
+      public void run() {
+        putSomeValues(500);
+      }
+    });
+
+    // Create the PR region
+    server2.invoke(SnapshotTxGIIDUnit.class, "createPR", new Object[]{regionName, 2, 1});
+
+    waitForGIICallbackStarted(-2, InitialImageOperation.GIITestHookType.BeforeRequestRVV);
+
+    unblockGII(-2, InitialImageOperation.GIITestHookType.BeforeRequestRVV);
+
+    // Wait for the PR to get initialized and GII completes on the lone bucket
+    waitForRegionInit(-2);
+
+    ai2.join();
+
+    RegionVersionVector rvv1 = getRVV(-1);
```

Comment:
Should we also check whether the rvv is same just after the gii is complete and before ai2 theoretically could have advanced because it would have been waiting on the readlock. so gii and gii with later puts both will be covered.